<HTML>
    <BODY>
        <H1 style="color:#e320c2">Distance de Mahalanobis</H1>
        <P>La distance de Mahalanobis est une métrique de distance multivariée efficace qui mesure la distance entre un point et une distribution. C'est une métrique extrêmement utile ayant d'excellentes applications dans la détection d'anomalies multivariées, la classification sur des ensembles de données très déséquilibrés et la classification à une classe. Cet article explique l'intuition et les mathématiques avec des exemples pratiques sur trois cas d'utilisation de l'apprentissage automatique.</P>
       </BODY>
</HTML>

<HTML>
    <BODY>
        <H1 style="color:#0409A0">1. Introduction</H1>
        <P>La distance de Mahalanobis est une métrique de distance multivariée efficace qui mesure la distance entre un point (vecteur) et une distribution .Il a d'excellentes applications dans la détection d'anomalies multivariées, la classification sur des ensembles de données très déséquilibrés et la classification à une classe et des cas d'utilisation plus inexploités.

Compte tenu de ses applications extrêmement utiles, cette métrique est rarement discutée ou utilisée dans les statistiques ou les flux de travail ML. Cet article explique le pourquoi et le quand utiliser la distance de Mahalanobis, puis explique l'intuition et les mathématiques avec des applications utiles.</P>
       </BODY>
</HTML>

<HTML>
    <BODY>
        <H1 style="color:#0409A0">2. Quel est le problème avec l'utilisation de la distance euclidienne pour les données multivariées ?</H1>
        <P>Commençons par les bases. La distance euclidienne est la distance en ligne droite couramment utilisée entre deux points.

Si les deux points sont dans un plan à deux dimensions (ce qui signifie que vous avez deux colonnes numériques (p) et (q)) dans votre ensemble de données), alors la distance euclidienne entre les deux points (p1, q1) et (p2, q2 ) est:</P>
        <img alt="image" src="img1.PNG">
        <P>Cette formule peut être étendue à autant de dimensions que vous le souhaitez :</P>  
        <img alt="image" src="img2.PNG">
        <p>Eh bien, la distance euclidienne fonctionnera bien tant que les dimensions sont pondérées de manière égale et indépendantes les unes des autres.

Qu'est-ce que je veux dire par là ?</p>
        <p>Considérons les tableaux suivants :</p>
        <img alt="image" src="img4.PNG">
        <P>Les deux tableaux ci-dessus indiquent la « superficie » et le « prix » des mêmes objets. Seules les unités des variables changent.

Étant donné que les deux tableaux représentent les mêmes entités, la distance entre deux lignes quelconques, le point A et le point B doit être la même. Mais la distance euclidienne donne une valeur différente même si les distances sont techniquement les mêmes dans l'espace physique.

Cela peut être techniquement surmonté en mettant les variables à l'échelle, en calculant le score z (ex : (x - moyen) / std) ou en le faisant varier dans une plage particulière comme entre 0 et 1.

Mais il y a un autre inconvénient majeur.

Autrement dit, si les dimensions (colonnes de votre jeu de données) sont corrélées les unes aux autres, ce qui est généralement le cas dans les jeux de données du monde réel, la distance euclidienne entre un point et le centre des points (distribution) peut donner des informations peu ou trompeuses à quel point un point est vraiment proche du cluster.</P>
         <img alt="image" src="img5.PNG">
        <p>L'image ci-dessus (à droite) est un simple nuage de points de deux variables positivement corrélées l'une à l'autre. Autrement dit, à mesure que la valeur d'une variable (axe des x) augmente, la valeur de l'autre variable (axe des y) augmente également.

Les deux points ci-dessus sont à égale distance (euclidien) du centre. Mais un seul d'entre eux (bleu) est en réalité plus proche de l'amas, même si, techniquement, la distance euclidienne entre les deux points est égale.
        En effet, la distance euclidienne est une distance entre deux points seulement. Il ne tient pas compte de la variation des autres points de l'ensemble de données. Ainsi, il ne peut pas être utilisé pour vraiment juger à quel point un point est réellement proche d'une distribution de points.

Ce dont nous avons besoin ici, c'est d'une métrique de distance plus robuste qui soit une représentation précise de la distance entre un point et une distribution.</p>      
   </BODY>
</HTML>

<HTML>
    <BODY>
        <H1 style="color:#0409A0">3. Quelle est la distance de Mahalanobis ?</H1>
        <p>La distance de Mahalonobis est la distance entre un point et une distribution. Et non entre deux points distincts. C'est effectivement un équivalent multivarié de la distance euclidienne.

Il a été introduit par le professeur P. C. Mahalanobis en 1936 et a depuis été utilisé dans diverses applications statistiques. Cependant, il n'est pas si bien connu ou utilisé dans la pratique de l'apprentissage automatique. Eh bien, entrons-y.
Donc, d'un point de vue informatique, en quoi la distance de Mahalanobis est-elle différente de la distance euclidienne ?</p>
        <p>-Il transforme les colonnes en variables non corrélées</p>
        <p>-Mettez les colonnes à l'échelle pour que leur variance soit égale à 1</p>
        <p>-Enfin, il calcule la distance euclidienne.</p>
<p>Les trois étapes ci-dessus visent à résoudre les problèmes de distance euclidienne dont nous venons de parler. Mais comment?

Regardons la formule et essayons de comprendre ses composants. </p>
   </BODY>
</HTML>

<HTML>
    <BODY>
        <H1 style="color:#0409A0">4.Les mathématiques et l'intuition derrière Mahalanobis Distance</H1>
        <P>La formule pour calculer la distance de Mahalanobis est la suivante :</P>
        <img alt="image" src="img3.PNG">
        <p>where, 
            - D^2 est le carré de la distance de Mahalanobis. </p>
        <p> - x est le vecteur de l'observation (ligne dans un jeu de données),</p> 
        <p> - m est le vecteur des valeurs moyennes des variables indépendantes (moyenne de chaque colonne),</p>
        <p> - C^(-1) est la matrice de covariance inverse des variables indépendantes.</p>
        
   </BODY>
</HTML>

<html>
    <body>
        <p>Prenons le (x – m)^T . C^(-1) terme.

(x – m) est essentiellement la distance du vecteur à la moyenne. Nous divisons ensuite cela par la matrice de covariance (ou multiplions par l'inverse de la matrice de covariance).

Si vous y réfléchissez, il s'agit essentiellement d'un équivalent multivarié de la normalisation régulière (z = (x – mu)/sigma). C'est-à-dire z = (vecteur x) – (vecteur moyen) / (matrice de covariance).

Alors, quel est l'effet de la division par la covariance ?

Si les variables de votre ensemble de données sont fortement corrélées, la covariance sera élevée. La division par une grande covariance réduira efficacement la distance.

De même, si les X ne sont pas corrélés, alors la covariance n'est pas élevée et la distance n'est pas beaucoup réduite.

Donc effectivement, il aborde à la fois les problèmes d'échelle ainsi que la corrélation des variables dont nous avons parlé dans l'introduction.</p>
    </body>
</html>
    

<html>
    <body>
     <p>Alors, comment comprendre la formule ci-dessus ?</p>
     <p>Prenons le (x – m)^T . C^(-1) terme.

(x – m) est essentiellement la distance du vecteur à la moyenne. Nous divisons ensuite cela par la matrice de covariance (ou multiplions par l'inverse de la matrice de covariance).

Si vous y réfléchissez, il s'agit essentiellement d'un équivalent multivarié de la normalisation régulière (z = (x – mu)/sigma). C'est-à-dire z = (vecteur x) – (vecteur moyen) / (matrice de covariance).

Alors, quel est l'effet de la division par la covariance ?

Si les variables de votre ensemble de données sont fortement corrélées, la covariance sera élevée. La division par une grande covariance réduira efficacement la distance.

De même, si les X ne sont pas corrélés, alors la covariance n'est pas élevée et la distance n'est pas beaucoup réduite.

Donc effectivement, il aborde à la fois les problèmes d'échelle ainsi que la corrélation des variables dont nous avons parlé dans l'introduction.</p>
    </body>
</html>
        

<HTML>
    <BODY>
        <H1 style="color:#0409A0">5.Comment calculer la distance de Mahalanobis en Python</H1>
   </BODY>
</HTML>

In [1]:
import pandas as pd
import scipy as sp
import numpy as np

filepath = 'https://raw.githubusercontent.com/selva86/datasets/master/diamonds.csv'
df = pd.read_csv(filepath).iloc[:, [0,4,6]]
df.head()

carat  depth  price
0   0.23   61.5    326
1   0.21   59.8    326
2   0.23   56.9    327
3   0.29   62.4    334
4   0.31   63.3    335

<HTML>
    <BODY>
        <p>Écrivons la fonction pour calculer la distance de Mahalanobis.</p>
    </BODY>
</HTML>    

In [5]:
import scipy as sp
def mahalanobis(x=None, data=None, cov=None):
    """Compute the Mahalanobis Distance between each row of x and the data  
    x    : vector or matrix of data with, say, p columns.
    data : ndarray of the distribution from which Mahalanobis distance of each observation of x is to be computed.
    cov  : covariance matrix (p x p) of the distribution. If None, will be computed from data.
    """
    x_minus_mu = x - np.mean(data)
    if not cov:
        cov = np.cov(data.values.T)
    inv_covmat = sp.linalg.inv(cov)
    left_term = np.dot(x_minus_mu, inv_covmat)
    mahal = np.dot(left_term, x_minus_mu.T)
    return mahal.diagonal()

df_x = df[['carat', 'depth', 'price']].head(500)
df_x['mahala'] = mahalanobis(x=df_x, data=df[['carat', 'depth', 'price']])
df_x.head()

carat  depth  price     mahala
0   0.23   61.5    326   1.709860
1   0.21   59.8    326   3.540097
2   0.23   56.9    327  12.715021
3   0.29   62.4    334   1.454469
4   0.31   63.3    335   2.347239

<html>
    <body>
        <h2><b style="color:#F17F0C">Cas d'utilisation 1 : détection des valeurs aberrantes multivariées à l'aide de la distance de Mahalanobis</b></h2>
        <p>En supposant que la statistique de test suit une distribution du chi carré avec « n » degrés de liberté, la valeur critique à un niveau de signification de 0,01 et 2 degrés de liberté est calculée comme suit :</p>
    </body>
 </html>   

In [3]:
# Valeurs critiques pour deux degrés de liberté
from scipy.stats import chi2
chi2.ppf((1-0.01), df=2)
#> 9.21

9.21034037197618

<html>
   <body>
       <p>Cela signifie qu'une observation peut être considérée comme extrême si sa distance de Mahalanobis dépasse 9,21.
Si vous préférez plutôt les valeurs P pour déterminer si une observation est extrême ou non, les valeurs P peuvent être calculées comme suit :</p>
    </body>
</html>    

In [6]:
# Compute the P-Values
df_x['p_value'] = 1 - chi2.cdf(df_x['mahala'], 2)

# Extreme values with a significance level of 0.01
df_x.loc[df_x.p_value < 0.01].head(10)

carat  depth  price     mahala   p_value
2     0.23   56.9    327  12.715021  0.001734
91    0.86   55.1   2757  23.909643  0.000006
97    0.96   66.3   2759  11.781773  0.002765
172   1.17   60.2   2774   9.279459  0.009660
204   0.98   67.9   2777  20.086616  0.000043
221   0.70   57.2   2782  10.405659  0.005501
227   0.84   55.1   2782  23.548379  0.000008
255   1.05   65.8   2789  11.237146  0.003630
284   1.00   58.2   2795  10.349019  0.005659
298   1.01   67.4   2797  17.716144  0.000142

<html>
    <body>
        <p>Si vous comparez les observations ci-dessus au reste de l'ensemble de données, elles sont clairement extrêmes.</p>
    </body>
 </html>   

<html>
    <body>
        <h2><b style="color:#F17F0C">Cas d'utilisation 2 : Distance de Mahalanobis pour les problèmes de classification</b></h2>
        <p>La distance Mahalanobis peut être utilisée pour des problèmes de classification. Une implémentation naïve d'un classificateur Mahalanobis est codée ci-dessous. L'intuition est qu'une observation se voit attribuer la classe dont elle est la plus proche en fonction de la distance de Mahalanobis.

Voyons un exemple d'implémentation sur le jeu de données BreastCancer, où l'objectif est de déterminer si une tumeur est bénigne ou maligne.</p>
    </body>
 </html>  

In [7]:
df = pd.read_csv('https://raw.githubusercontent.com/selva86/datasets/master/BreastCancer.csv', 
                 usecols=['Cl.thickness', 'Cell.size', 'Marg.adhesion', 
                          'Epith.c.size', 'Bare.nuclei', 'Bl.cromatin', 'Normal.nucleoli', 
                          'Mitoses', 'Class'])

df.dropna(inplace=True)  # drop missing values.
df.head()

Cl.thickness  Cell.size  Marg.adhesion  Epith.c.size  Bare.nuclei  \
0             5          1              1             2          1.0   
1             5          4              5             7         10.0   
2             3          1              1             2          2.0   
3             6          8              1             3          4.0   
4             4          1              3             2          1.0   

   Bl.cromatin  Normal.nucleoli  Mitoses  Class  
0            3                1        1      0  
1            3                2        1      0  
2            3                1        1      0  
3            3                7        1      0  
4            3                1        1      0

<html>
    <body>
        <p>Séparons l'ensemble de données dans un rapport 70:30 en tant qu'entraînement et test. Et l'ensemble de données d'apprentissage est divisé en groupes homogènes de classes 'pos'(1) et 'neg'(0). Pour prédire la classe de l'ensemble de données de test, nous mesurons les distances de Mahalanobis entre une observation donnée (ligne) et les ensembles de données positifs (xtrain_pos) et négatifs (xtrain_neg).

Ensuite, cette observation est attribuée à la classe en fonction du groupe dont elle est la plus proche.</p>
    </body>
 </html>   

In [8]:
from sklearn.model_selection import train_test_split
xtrain, xtest, ytrain, ytest = train_test_split(df.drop('Class', axis=1), df['Class'], test_size=.3, random_state=100)

# Split the training data as pos and neg
xtrain_pos = xtrain.loc[ytrain == 1, :]
xtrain_neg = xtrain.loc[ytrain == 0, :]

<html>
    <body>
        <p>Construisons le MahalanobiBinaryClassifier. Pour ce faire, vous devez définir les méthodes predict_proba() et predict(). Ce classificateur ne nécessite pas de méthode fit() (entraînement) distincte.</p>
    </body>
 </html>   

In [9]:
class MahalanobisBinaryClassifier():
    def __init__(self, xtrain, ytrain):
        self.xtrain_pos = xtrain.loc[ytrain == 1, :]
        self.xtrain_neg = xtrain.loc[ytrain == 0, :]

    def predict_proba(self, xtest):
        pos_neg_dists = [(p,n) for p, n in zip(mahalanobis(xtest, self.xtrain_pos), mahalanobis(xtest, self.xtrain_neg))]
        return np.array([(1-n/(p+n), 1-p/(p+n)) for p,n in pos_neg_dists])

    def predict(self, xtest):
        return np.array([np.argmax(row) for row in self.predict_proba(xtest)])


clf = MahalanobisBinaryClassifier(xtrain, ytrain)        
pred_probs = clf.predict_proba(xtest)
pred_class = clf.predict(xtest)

# Pred and Truth
pred_actuals = pd.DataFrame([(pred, act) for pred, act in zip(pred_class, ytest)], columns=['pred', 'true'])
print(pred_actuals[:5])     

   pred  true
0     0     0
1     1     1
2     0     0
3     0     0
4     0     0


<html>
    <body>
        <p>Voyons comment le classificateur a fonctionné sur l'ensemble de données de test.</p>
    </body>
 </html>   

In [10]:
from sklearn.metrics import classification_report, accuracy_score, roc_auc_score, confusion_matrix
truth = pred_actuals.loc[:, 'true']
pred = pred_actuals.loc[:, 'pred']
scores = np.array(pred_probs)[:, 1]
print('AUROC: ', roc_auc_score(truth, scores))
print('\nConfusion Matrix: \n', confusion_matrix(truth, pred))
print('\nAccuracy Score: ', accuracy_score(truth, pred))
print('\nClassification Report: \n', classification_report(truth, pred))

AUROC:  0.990974358974359

Confusion Matrix: 
 [[113  17]
 [  0  75]]

Accuracy Score:  0.9170731707317074

Classification Report: 
               precision    recall  f1-score   support

           0       1.00      0.87      0.93       130
           1       0.82      1.00      0.90        75

    accuracy                           0.92       205
   macro avg       0.91      0.93      0.91       205
weighted avg       0.93      0.92      0.92       205



<html>
    <body>
        <p>La distance de Mahalanobis à elle seule peut contribuer à cette précision (92%).</p>
    </body>
 </html>   



<html>
    <body>
        <h2><b style="color:#F17F0C">Cas d'utilisation 3 : Classification à une classe</b></h2>
        <p>La classification à une classe est un type d'algorithme dans lequel l'ensemble de données d'apprentissage contient des observations appartenant à une seule classe.

Avec seulement cette information connue, l'objectif est de déterminer si une observation donnée dans un nouvel ensemble de données (ou de test) appartient à cette classe.

Vous pourriez vous demander quand une telle situation se produirait. Eh bien, c'est un problème assez courant en Data Science.

Par exemple, considérons la situation suivante : vous disposez d'un vaste ensemble de données contenant des millions d'enregistrements qui ne sont PAS encore classés en 1 et en 0. Mais vous avez également avec vous un petit échantillon de données contenant uniquement des enregistrements positifs. En apprenant les informations de cet exemple de jeu de données, vous souhaitez classer tous les enregistrements du grand jeu de données en 1 et en 0.

Sur la base des informations de l'ensemble de données de l'échantillon, il est possible de dire si un échantillon donné est un 1 ou un 0 en ne visualisant que les 1 (et en n'ayant aucune connaissance des 0).

Cela peut être fait en utilisant la distance de Mahalanobis.

Essayons ceci sur l'ensemble de données BreastCancer, mais cette fois-ci, nous ne considérerons que les observations malignes (colonne de classe=1) dans les données d'entraînement.</p>
    </body>
 </html> 

In [11]:
df = pd.read_csv('https://raw.githubusercontent.com/selva86/datasets/master/BreastCancer.csv', 
                 usecols=['Cl.thickness', 'Cell.size', 'Marg.adhesion', 
                          'Epith.c.size', 'Bare.nuclei', 'Bl.cromatin', 'Normal.nucleoli', 
                          'Mitoses', 'Class'])

df.dropna(inplace=True)  # drop missing values.

<html>
    <body>
        <p>Diviser 50 % de l'ensemble de données en entraînement et test. Seuls les 1 sont conservés dans les données d'entraînement.</p>
    </body>
 </html>  


In [12]:
from sklearn.model_selection import train_test_split
xtrain, xtest, ytrain, ytest = train_test_split(df.drop('Class', axis=1), df['Class'], test_size=.5, random_state=100)

# Split the training data as pos and neg
xtrain_pos = xtrain.loc[ytrain == 1, :]

<html>
    <body>
        <p>Construisons le MahalanobisOneClassClassifier et obtenons la distance mahalanobis de chaque point de données en x à partir de l'ensemble d'apprentissage (xtrain_pos).</p>
    </body>
 </html>  



In [13]:
class MahalanobisOneclassClassifier():
    def __init__(self, xtrain, significance_level=0.01):
        self.xtrain = xtrain
        self.critical_value = chi2.ppf((1-significance_level), df=xtrain.shape[1]-1)
        print('Critical value is: ', self.critical_value)

    def predict_proba(self, xtest):
        mahalanobis_dist = mahalanobis(xtest, self.xtrain)
        self.pvalues = 1 - chi2.cdf(mahalanobis_dist, 2)
        return mahalanobis_dist

    def predict(self, xtest):
        return np.array([int(i) for i in self.predict_proba(xtest) > self.critical_value])

clf = MahalanobisOneclassClassifier(xtrain_pos, significance_level=0.05)
mahalanobis_dist = clf.predict_proba(xtest)

# Pred and Truth
mdist_actuals = pd.DataFrame([(m, act) for m, act in zip(mahalanobis_dist, ytest)], columns=['mahal', 'true_class'])
print(mdist_actuals[:5])  

Critical value is:  14.067140449340169
       mahal  true_class
0  13.104716           0
1  14.408570           1
2  14.932236           0
3  14.588622           0
4  15.471064           0


<html>
    <body>
        <p>Nous avons la distance de Mahalanobis et la classe réelle de chaque observation.

Je m'attendrais à ce que ces observations avec une faible distance de Mahalanobis soient de 1.

Donc, je trie les mdist_actuals par distance de Mahalanobis et le quantile coupe les lignes en 10 groupes de taille égale. Les observations dans les quantiles supérieurs devraient avoir plus de 1 par rapport à celles du bas. Voyons.</p>
    </body>
 </html>  


In [14]:
# quantile cut in 10 pieces
mdist_actuals['quantile'] = pd.qcut(mdist_actuals['mahal'], 
                                    q=[0, .10, .20, .3, .4, .5, .6, .7, .8, .9, 1], 
                                    labels=[1, 2, 3, 4, 5, 6, 7, 8, 9, 10])

# sort by mahalanobis distance
mdist_actuals.sort_values('mahal', inplace=True)
perc_truths = mdist_actuals.groupby('quantile').agg({'mahal': np.mean, 'true_class': np.sum}).rename(columns={'mahal':'avg_mahaldist', 'true_class':'sum_of_trueclass'})
print(perc_truths)

          avg_mahaldist  sum_of_trueclass
quantile                                 
1              3.765496                33
2              6.511026                32
3              9.272944                30
4             12.209504                20
5             14.455050                 4
6             15.684493                 4
7             17.368633                 3
8             18.840714                 1
9             21.533159                 2
10            23.524055                 1


<html>
    <body>
        <p>Si vous remarquez ci-dessus, près de 90 % des 1 (cas malins) se situent dans les premiers 40 % de la distance de Mahalanobis. Incidemment, tous ces éléments sont inférieurs à la valeur critique pf 14,05. Alors, prenons la valeur critique comme seuil et marquons ces observations avec une distance de Mahalanobis inférieure au seuil comme positives.</p>
    </body>
 </html>  


In [15]:
from sklearn.metrics import classification_report, accuracy_score, roc_auc_score, confusion_matrix

# Positive if mahalanobis 
pred_actuals = pd.DataFrame([(int(p), y) for y, p in zip(ytest, clf.predict_proba(xtest) < clf.critical_value)], columns=['pred', 'true'])

# Accuracy Metrics
truth = pred_actuals.loc[:, 'true']
pred = pred_actuals.loc[:, 'pred']
print('\nConfusion Matrix: \n', confusion_matrix(truth, pred))
print('\nAccuracy Score: ', accuracy_score(truth, pred))
print('\nClassification Report: \n', classification_report(truth, pred))


Confusion Matrix: 
 [[183  29]
 [ 15 115]]

Accuracy Score:  0.8713450292397661

Classification Report: 
               precision    recall  f1-score   support

           0       0.92      0.86      0.89       212
           1       0.80      0.88      0.84       130

    accuracy                           0.87       342
   macro avg       0.86      0.87      0.87       342
weighted avg       0.88      0.87      0.87       342



<html>
    <body>
        <p>Ainsi, sans la connaissance de la classe bénigne, nous sommes en mesure de prédire avec précision la classe de 87 % des observations</p>
    </body>
 </html> 


<HTML>
    <BODY>
        <H1 style="color:#0409A0"> Conclusion</H1>
        <P>La distance de Mahalanobis est une métrique de distance multivariée efficace qui mesure la distance entre un point (vecteur) et une distribution .Il a d'excellentes applications dans la détection d'anomalies multivariées, la classification sur des ensembles de données très déséquilibrés et la classification à une classe et des cas d'utilisation plus inexploités.

Compte tenu de ses applications extrêmement utiles, cette métrique est rarement discutée ou utilisée dans les statistiques ou les flux de travail ML. Cet article explique le pourquoi et le quand utiliser la distance de Mahalanobis, puis explique l'intuition et les mathématiques avec des applications utiles.</P>
    </BODY>
</HTML>    
